src/style/fog.js

```diff
@@ -29,16 +34,77 @@ const TRANSITION_SUFFIX = '-transition';
 export const FOG_PITCH_START = 55;
 export const FOG_PITCH_END = 65;
 
+export class FogSampler {
+    properties: ?PossiblyEvaluated<Props>;
+
+    // As defined in _prelude_fog.fragment.glsl#fog_opacity
+    getFogOpacity(depth: number, pitch: number): number {
+        if (!this.properties) { return 0.0; }
+
+        const props = this.properties;
+        const range = props.get('range');
+        const fogOpacity = props.get('opacity') * smoothstep(FOG_PITCH_START, FOG_PITCH_END, pitch);
+        const start = range[0], end = range[1];
+
+        // The fog is not physically accurate, so we seek an expression which satisfies a
+        // couple basic constraints:
+        //   - opacity should be 0 at the near limit
+        //   - opacity should be 1 at the far limit
+        //   - the onset should have smooth derivatives to avoid a sharp band
+        // To this end, we use an (1 - e^x)^n, where n is set to 3 to ensure the
+        // function is C2 continuous at the onset. The fog is about 99% opaque at
+        // the far limit, so we simply scale it and clip to achieve 100% opacity.
+        // https://www.desmos.com/calculator/3taufutxid
```

Comment:
Maybe styles replicating shaders is standard practice, but perhaps a reference, just in case something were to change?

```suggestion
        // https://www.desmos.com/calculator/3taufutxid
        // The output of this function must match src/shaders/_prelude_fog.fragment.glsl
```